pkg/providers/goinstall.go

```diff
@@ -0,0 +1,126 @@
+package providers
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/apex/log"
+)
+
+type goinstall struct {
+	name, repo, tag, latestURL string
+}
+
+func parseRepo(path string) (string, string, string, string) {
+	repo := path
+	tag := "latest"
+	if i := strings.LastIndex(path, "@"); i > -1 {
+		repo = filepath.Clean(path[:i])
+		tag = path[i+1:]
+	}
+
+	name := path
+	if i := strings.LastIndex(repo, "/"); i > -1 {
+		name = repo[i+1:]
+	}
+
+	latestURL := fmt.Sprintf("https://proxy.golang.org/%s/@latest", repo)
+
+	return repo, tag, name, latestURL
+}
+
+func newGoInstall(repo string) (Provider, error) {
+	repoUrl := strings.TrimPrefix(repo, "goinstall://")
+	repo, tag, name, latestURL := parseRepo(repoUrl)
+	return &goinstall{repo: repo, tag: tag, name: name, latestURL: latestURL}, nil
+}
+
+func (g *goinstall) Fetch(opts *FetchOpts) (*File, error) {
+
+	if (len(g.tag) > 0 && g.tag != "latest") || len(opts.Version) > 0 {
+		if len(opts.Version) > 0 {
+			// this is used by for the `ensure` command
+			g.tag = opts.Version
+		}
+		log.Infof("Getting %s release for %s", g.tag, g.repo)
+	} else {
+		log.Infof("Getting latest release for %s", g.repo)
+		if name, _, err := g.GetLatestVersion(); err != nil {
+			return nil, fmt.Errorf("failed to get latest version: %w", err)
+		} else {
+			g.tag = name
+		}
+	}
+
+	cmd := exec.Command("go", "install", fmt.Sprintf("%s@%s", g.repo, g.tag))
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("failed to install package: %w", err)
+	}
+
+	goBinPath := fmt.Sprintf("%s/bin/%s", os.Getenv("GOPATH"), g.name)
+
+	file, err := os.Open(os.ExpandEnv(goBinPath))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var buf bytes.Buffer
+	_, err = io.Copy(&buf, file)
```

Comment:
do we need to read the whole file in memory here? Why not just assign the `file` to `Data` directly? Since `Data` expects an `io.Reader` which `file` already is.

Reply:
I use to always close a file when i open it plus after i delete it, so i can't use it in return.

Reply:
Not sure that it will be cross-platform, because Linux systems use reference counting for inodes where Windows lock and prevents deletion.
So to be sure it's cross-platform i kould keep the copy or make a difference using runtime.GOOS, if windows copy else  return file ?